src/inventorymanager.cpp

```diff
@@ -154,6 +154,67 @@ IMoveAction::IMoveAction(std::istream &is, bool somewhere) :
 	}
 }
 
+void IMoveAction::swapDirections()
+{
+	std::swap(from_inv, to_inv);
+	std::swap(from_list, to_list);
+	std::swap(from_i, to_i);
+}
+
+void IMoveAction::onPutAndOnTake(ItemStack &src_item, ServerActiveObject* player)
+{
+	if (from_inv.type == InventoryLocation::DETACHED)
+		PLAYER_TO_SA(player)->detached_inventory_OnPut(
+			*this, src_item, player);
+	else if (from_inv.type == InventoryLocation::NODEMETA)
+		PLAYER_TO_SA(player)->nodemeta_inventory_OnPut(
+			*this, src_item, player);
+	else if (from_inv.type == InventoryLocation::PLAYER)
+		PLAYER_TO_SA(player)->player_inventory_OnPut(
+			*this, src_item, player);
+	if (from_inv.type == InventoryLocation::DETACHED)
+		PLAYER_TO_SA(player)->detached_inventory_OnTake(
+			*this, src_item, player);
+	else if (from_inv.type == InventoryLocation::NODEMETA)
+		PLAYER_TO_SA(player)->nodemeta_inventory_OnTake(
+			*this, src_item, player);
+	else if (from_inv.type == InventoryLocation::PLAYER)
+		PLAYER_TO_SA(player)->player_inventory_OnTake(
+			*this, src_item, player);
+}
+
+int IMoveAction::allowPut(ItemStack &dst_item, ServerActiveObject* player)
+{
+	int dst_can_put_count = 0xffff;
+	if (to_inv.type == InventoryLocation::DETACHED) {
+		dst_can_put_count = PLAYER_TO_SA(player)->detached_inventory_AllowPut(
+			*this, dst_item, player);
+	} else if (to_inv.type == InventoryLocation::NODEMETA) {
+		dst_can_put_count = PLAYER_TO_SA(player)->nodemeta_inventory_AllowPut(
+			*this, dst_item, player);
+	} else if (to_inv.type == InventoryLocation::PLAYER) {
+		dst_can_put_count = PLAYER_TO_SA(player)->player_inventory_AllowPut(
+			*this, dst_item, player);
+	}
```

Comment:
If this is supposed to be exhaustive please add an `assert(false)` in an else branch
If not then the behaviour is subtly different than before and this has to be looked at.